common/util/PerfCounter.java

```diff
@@ -0,0 +1,42 @@
+package com.vaticle.typedb.core.common.util;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+
+public abstract class PerfCounter {
+
+    public static PerfCounterImpl createCounter() { return new PerfCounterImpl(); }
+
+    public static NoOpPerfCounter createNoOpCounter() { return new NoOpPerfCounter(); }
+
+    public abstract long get(Object identifier);
+
+    public abstract void add(Object identifier, long delta);
+
+    private static class PerfCounterImpl extends PerfCounter {
+
+        private final Map<Object, AtomicLong> counters;
+
+        public PerfCounterImpl() {
+            counters = new HashMap<>();
```

Comment:
this needs to be a `ConcurrentHashMap`, and then you can use normal non-atomic longs since it's protected by the concurrent hash map protection

Reply:
Why would you use a ConcurrentHashMap when you can use atomic operations?